### DISCUSSION TITLE
Getting user commit count for the month

### DISCUSSION THREAD

**Discussion by AFCMS** [community/community#114470] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Basically I want to get the value here I can see as an authenticated user:

![image](https://github.com/community/community/assets/61794590/e7e4be62-3965-4284-aa2b-e82893e43650)

I have tried the following code running on Vercel:

```typescript
import type {VercelRequest, VercelResponse} from "@vercel/node";
import {Octokit} from "octokit";

interface PushEvent {
    type: string;
    payload: {
        size: number;
    };
}

// noinspection JSUnusedGlobalSymbols
export default async function handler(req: VercelRequest, res: VercelResponse) {
    const GITHUB_API_TOKEN = process.env.GITHUB_API_TOKEN
    const octokit = new Octokit({auth: GITHUB_API_TOKEN})

    const oneWeekAgo = new Date();
    oneWeekAgo.setDate(oneWeekAgo.getDate() - 7);

    const startOfMarch = new Date();
    startOfMarch.setDate(1);
    startOfMarch.setHours(0, 0, 0, 0);

    const endOfMarch = new Date();
    endOfMarch.setDate(31);
    endOfMarch.setHours(23, 59, 59, 999);

    let page = 1;
    let allEvents = [];

    while (true) {
        const events = await octokit.rest.activity.listEventsForAuthenticatedUser({
            username: "AFCMS",
            page: page,
            per_page: 100,
        });

        allEvents = allEvents.concat(events.data);
        page++;

        if (events.data.length === 0 || new Date(events.data[events.data.length - 1].created_at) < startOfMarch) {
            break;
        }
    }

    console.log(allEvents);
    console.log(new Date(allEvents[allEvents.length - 1].created_at).toLocaleDateString());
    console.log("LENGHT:", allEvents.length);


    const userEvents = allEvents.filter(event => new Date(event.created_at) >= startOfMarch && new Date(event.created_at) <= endOfMarch);

    console.log("FILTERED LENGHT:", userEvents.length);

    const pushEvents = userEvents.filter(event => event.type === "PushEvent");

    console.log("PUSH EVENTS:", pushEvents.length);

    const commitCount = pushEvents.reduce<number>((acc, event) => {
            return acc + event.payload.commits.filter(commit => commit.author.name === "AFCMS" && commit.author.email === "afcm.contact@gmail.com").length;
        }
    }, 0);
    //const pullRequestEvents = weeklyEvents.filter(event => event.type === "PullRequestEvent");

    console.log("COMMIT COUNT:", commitCount);

    const totalCommits = pushEvents.reduce((total, event) => total + (event as unknown as PushEvent).payload.size, 0);

    /*return res.json({
        totalCommits: totalCommits,
        //commits: allEvents,
    })*/

    return res.json(pushEvents)
}
```

The output commit count is ~176 which doesn't match what I see in my profile (while being authenticated ofc). I have tried filtering commits by email, but I am getting the same number.

I tried limiting to the base branches `master`/`main` using the following:

```typescript
const commitCount = pushEvents.reduce<number>((acc, event) => {
   if (event.payload.ref === "refs/heads/master" || event.payload.ref === "refs/heads/main") {
       return acc + event.payload.commits.filter(commit => commit.author.name === "AFCMS" && commit.author.email === "afcm.contact@gmail.com").length;
    }
    return acc;
}, 0);
```

But it outputs 113 which is also wrong

Any ideas?

**Comment by mapluisch** [comment]
Hi AFCMS,

I've changed your script quite a bit and made my own .js out of it (which works fine on my end):

```javascript
const { Octokit } = require("@octokit/rest");

module.exports = async (req, res) => {
    const GITHUB_TOKEN = process.env.GITHUB_API_TOKEN;
    const octokit = new Octokit({ auth: GITHUB_TOKEN });

    try {
        const repos = await octokit.paginate(octokit.repos.listForAuthenticatedUser, {
            per_page: 100,
        });

        let totalCommits = 0;

        for (const repo of repos) {
            const commits = await octokit.paginate(octokit.repos.listCommits, {
                owner: repo.owner.login,
                repo: repo.name,
                since: `2024-03-01T00:00:00Z`,
                until: `2024-03-31T23:59:59Z`,
            });
            
            totalCommits += commits.length;
        }

        res.json({ totalCommits });
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: "error fetching commit data from GitHub" });
    }
};
```

I'm using `octokit.repos.listCommits` to iterate through all commits and sum them up this way. I've also removed the way you constructed the start and end of march as you can make it a bit more streamlined (you could also adapt this code more easily to apply it for different months eg as param).

Let me know if it works :)

**Reply by AFCMS** [reply]
I wasn't aware of the paginate method! It's really nice to improve readability of the code.

Sadly the value is greater than the one I see on my profile (really strange IMO), still not what I expected and I would assume the commits in my organisations are not counted?

Also I have about 100 repositories so the request takes really long to complete (the Vercel function should be exploited in a IoT application so not really a good solution in this case).

Anyways thanks for the help!

EDIT: The easiest solution might be to do web scrapping to extract the data from the profile page directly :smile:

**Reply by mapluisch** [reply]
Sure, you're welcome - and sorry to hear it doesn't quite work just yet 😅

Yep, web scraping will be pretty straightforward and will show you the correct data, but you could run into rate-limits pretty easily (depending on how often you intend to fetch your data of course).

Good luck and happy coding :)

**Comment by LangLangBart** [comment]
> Basically I want to get the value here I can see as an authenticated user:
> 
> ![image](https://private-user-images.githubusercontent.com/61794590/317232248-e7e4be62-3965-4284-aa2b-e82893e43650.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxMzUyMzIsIm5iZiI6MTcxMjEzNDkzMiwicGF0aCI6Ii82MTc5NDU5MC8zMTcyMzIyNDgtZTdlNGJlNjItMzk2NS00Mjg0LWFhMmItZTgyODkzZTQzNjUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAzVDA5MDIxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiY2M1N2E4ZTUwZDc1YTI1OTQ5MjA3NzZmYzU5NGJjZTYxYjE5ZmJjMGM2NDI2MDA3OWQ2MmMxYzUwOWYyMWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.q4NzAcCNjbaTFaudUVgR9B7uBbpErbQCcgJ6S8Is0VQ)
> 
> I have tried the following code running on Vercel:


Instead of the `REST API`, consider using the `GraphQL API` with the `contributionsCollection`[^1] field.

The features of both APIs overlap but are not identical; think of them as a Venn diagram.

Refer to the exemplary `bun` script below for further clarity.


```ts
#!/usr/bin/env bun
import process from 'node:process'
import { Octokit } from 'octokit'
import type { GraphQlQueryResponseData } from '@octokit/graphql'
import { GraphqlResponseError } from '@octokit/graphql'

const contributionsCollection = `
query ($userName: String!, $startTime: DateTime, $endTime: DateTime) {
	user(login: $userName) {
		contributionsCollection(from: $startTime, to: $endTime) {
			totalCommitContributions
		}
	}
}
`

try {
	// Your token.
	const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN })
	// Use @sindresorhus as the default.
	const userName: string = process.argv[2] ?? 'sindresorhus'

	const now = new Date()
	// Generate ISO-8601 UTC date strings, e.g., "2024-04-01T00:00:00.000Z" in case of no input.
	const startTime = process.argv[3]
		?? new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1)).toISOString()
	const endTime = process.argv[4]
		?? new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth() + 1, 0, 23, 59, 59, 999)).toISOString()

	const response: GraphQlQueryResponseData = await octokit.graphql(contributionsCollection, {
		userName, startTime, endTime
	})
	console.log('Start Time:', startTime)
	console.log('End Time:', endTime)
	console.log(`Commits [${userName}]:`, response.user.contributionsCollection.totalCommitContributions)
}

catch (error) {
	if (error instanceof GraphqlResponseError) {
		console.error(`Failed to fetch contributionsCollection: ${error}`)
		throw error
	}
	else {
		console.error('Something went wrong…')
		throw error
	}
}
```

```bash
# Check sindresorhus's total commit contributions for the current month.
./octokit-contributions.ts
# Start Time: 2024-04-01T00:00:00.000Z
# End Time: 2024-04-30T23:59:59.999Z
# Commits [sindresorhus]: 31
```

```bash
# Check AFCMS's total commit contributions for March '24.
./octokit-contributions.ts 'AFCMS' 2024-03-01T00:00:00.000Z 2024-03-31T23:59:59.999Z
# Start Time: 2024-03-01T00:00:00.000Z
# End Time: 2024-03-31T23:59:59.999Z
# Commits [AFCMS]: 139
```

[^1]: [Objects - GitHub Docs GraphQL - contributionsCollection](https://docs.github.com/en/graphql/reference/objects#contributionscollection)

**Reply by AFCMS** [reply]
It seems to accurately return the public commit count, that's a big step forward actually! :sweat_smile:

But I would like to get the commit count you see in the authenticated profile (not specific user specified, use the one who owns the token). Do you know how you could get this with the GraphQL API? I never used GraphQL before.

EDIT: I replaced `user(login: $userName)` by `viewer` and it uses the token owner data, but it's still the public commit count :cry: 

EDIT2: The token have all repo permissions

**Reply by LangLangBart** [reply, marked as answer]
> EDIT: I replaced `user(login: $userName)` by `viewer` and it uses the token owner data

👍

> but it's still the public commit count 😢


I cannot reproduce your problem; the script returns the same number of commits that I see when I'm logged in.

When executing the script as demonstrated below, do you continue to receive 139 commits instead of 141 for March '24?

---

Modified version making a `viewer`[^1] query instead of a `user`[^2] query in `GraphQL`.

```ts
#!/usr/bin/env bun
import process from 'node:process'
import { Octokit } from 'octokit'
import type { GraphQlQueryResponseData } from '@octokit/graphql'
import { GraphqlResponseError } from '@octokit/graphql'

const contributionsCollection = `
query ($startTime: DateTime, $endTime: DateTime) {
	viewer {
		contributionsCollection(from: $startTime, to: $endTime) {
			totalCommitContributions
		}
	}
}
`

try {
	// Your token.
	const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN })

	const now = new Date()
	// Generate ISO-8601 UTC date strings, e.g., "2024-04-01T00:00:00.000Z" in case of no input.
	const startTime = process.argv[2]
		?? new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1)).toISOString()
	const endTime = process.argv[3]
		?? new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth() + 1, 0, 23, 59, 59, 999)).toISOString()

	const response: GraphQlQueryResponseData = await octokit.graphql(contributionsCollection, {
		startTime, endTime
	})
	console.log('Start Time:', startTime)
	console.log('End Time:', endTime)
	console.log('Your Commits:', response.viewer.contributionsCollection.totalCommitContributions)
}

catch (error) {
	if (error instanceof GraphqlResponseError) {
		console.error(`Failed to fetch contributionsCollection: ${error}`)
		throw error
	}
	else {
		console.error('Something went wrong…')
		throw error
	}
}
```




[^1]: [Queries: viewer - GitHub Docs GraphQL](https://docs.github.com/en/graphql/reference/queries#viewer)
[^2]: [Queries: user - GitHub Docs GraphQL](https://docs.github.com/en/graphql/reference/queries#user)

**Reply by AFCMS** [reply]
Was busy replying it didn't work but oh my god I figured it out

My token had the `repo` permission but not the `user` one :sweat_smile: 

In fact I even asked GitHub Copilot how to query the GraphQL API to include the private contribution and he replied to use the `repo` permission but not the `user` permission lol

Now I have the right amount of commits :partying_face: 

Thank you so much for your help!
